Sources/xcproj/PBXProj.swift

```diff
@@ -3,8 +3,135 @@ import PathKit
+                lhs.nativeTargets == rhs.nativeTargets &&
+                lhs.fileReferences == rhs.fileReferences &&
+                lhs.projects == rhs.projects &&
+                lhs.versionGroups == rhs.versionGroups &&
+                lhs.referenceProxies == rhs.referenceProxies
+        }
+
+        // MARK: - Public Methods
+
+        public func addObject(_ object: PBXObject) {
+            switch object {
+            case let object as PBXBuildFile: buildFiles.append(object)
+            case let object as PBXAggregateTarget: aggregateTargets.append(object)
+            case let object as PBXContainerItemProxy: containerItemProxies.append(object)
+            case let object as PBXCopyFilesBuildPhase: copyFilesBuildPhases.append(object)
+            case let object as PBXGroup: groups.append(object)
+            case let object as PBXFileElement: fileElements.append(object)
+            case let object as XCConfigurationList: configurationLists.append(object)
+            case let object as XCBuildConfiguration: buildConfigurations.append(object)
+            case let object as PBXVariantGroup: variantGroups.append(object)
+            case let object as PBXTargetDependency: targetDependencies.append(object)
+            case let object as PBXSourcesBuildPhase: sourcesBuildPhases.append(object)
+            case let object as PBXShellScriptBuildPhase: shellScriptBuildPhases.append(object)
+            case let object as PBXResourcesBuildPhase: resourcesBuildPhases.append(object)
+            case let object as PBXFrameworksBuildPhase: frameworksBuildPhases.append(object)
+            case let object as PBXHeadersBuildPhase: headersBuildPhases.append(object)
+            case let object as PBXNativeTarget: nativeTargets.append(object)
+            case let object as PBXFileReference: fileReferences.append(object)
+            case let object as PBXProject: projects.append(object)
+            case let object as XCVersionGroup: versionGroups.append(object)
+            case let object as PBXReferenceProxy: referenceProxies.append(object)
+            default: fatalError("Unhandled PBXObject type for \(object), this is likely a bug / todo")
+            }
+        }
+
+        public func getReference<T: PBXObject>(_ reference: String) -> T? {
+            let caches: [[String: PBXObject]] = [
+                buildFiles,
+                aggregateTargets,
+                containerItemProxies,
+                groups,
+                fileElements,
+                configurationLists,
+                buildConfigurations,
+                variantGroups,
+                targetDependencies,
+                nativeTargets,
+                fileReferences,
+                projects,
+                versionGroups,
+                referenceProxies
+            ]
+            return caches.first { cache in cache[reference] != nil }?[reference] as? T ?? self.buildPhases.first { $0.reference == reference } as? T ?? nil
```

Comment:
After moving the buildPhases to dictionaries you'll be able to simplify this line